### ISSUE TITLE
False report CA1821 (Remove empty Finalizers)

### ISSUE THREAD

**Issue by mavasani** [dotnet/roslyn-analyzers#1241]
See https://github.com/dotnet/roslyn/pull/19806/files#diff-4fddad076e8fb11cecceca523ee2aeb5R120

**Comment by mavasani** [comment]
@nguerrera Thanks, you are right that the analyzer is reporting a valid issue on release build. However, there is still an issue in the analyzer - it shouldn't fire if the enclosing method is also conditionally excluded. For example, the following still fires the diagnostic in both release and debug builds.

#if DEBUG
        ~InvisibleEditor()
        {
            Debug.Assert(Environment.HasShutdownStarted, GetType().Name + " was leaked without Dispose being called.");
        }
#endif

**Comment by sharwell** [comment]
> For example, the following still fires the diagnostic in both release and debug builds.

How does the diagnostic fire in this case for release builds?

**Comment by 333fred** [comment]
The problem here is that we explicitly flag bodies with just methods marked as conditional, but we don't then check to see if the method is excluded conditionally. I've filed https://github.com/dotnet/roslyn/issues/24569 for exposing the ability to get preprocessor info for a given syntax node, so we can check to see if the method is guarded by `#if DEBUG` flags.

**Comment by sharwell** [comment]
> The problem here is that we explicitly flag bodies with just methods marked as conditional, but we don't then check to see if the method is excluded conditionally.

My understanding is you currently treat all conditional methods as not included in the compilation output. After the change, you will start to treat them as included if the conditional symbol applied to the method definition is defined at the point of the call. Is this correct?

**Comment by 333fred** [comment]
Yep, that's correct. The idea is that if the user has a method with just `Debug` calls in it, we can then tell them to wrap with `#if DEBUG`.

**Comment by Cyberboss** [comment]
Here is another repro: https://github.com/tgstation/tgstation-server/blob/master/src/Tgstation.Server.Host/Components/Watchdog/SessionController.cs#L230

**Comment by Evangelink** [comment]
@mavasani We could use a workaround while waiting for a more reliable solution. My idea is to count invocations to method with the conditional attribute if the destructor has any directive.

This would result in potentially some false negative but would certainly kill the false positives raised here. Besides, I think that the current implementation might be a little to strict anyways because a method marked with `ConditionalAttribute` specifies which pre-processor symbol is expected while our current implementation rules them all out. WDYT?

**Comment by sharwell** [comment]
@Evangelink the methods linked by @Cyberboss do not involve preprocessor symbols. Can you clarify what your proposed change is, and which scenarios it would address?

@Cyberboss Both of your finalizers involve invalid calls to other managed objects. The most straightforward way to address this problem is to remove both of the finalizers.

1. `SessionController.Dispose(bool)` should not be calling `process.Terminate` or `logger.LogCritical` when `disposing` is false. Since these are the only significant calls made on the finalization path, the finalizer should be removed.
2. `SemaphoreSlimContext.Dispose()` should not be calling `lockedSemaphore.Release` on the finalization path. Since this is the only significant call made on the finalization path, the finalizer should be removed.

Depending on how your code is structured, it's possible that your application is currently relying on these finalizers for some normal scenarios. Even though they are working in some cases, they are generally not guaranteed to work under any particular scenario so it would be good to remove them.

**Comment by Evangelink** [comment]
@sharwell I was talking about the case where the finalizer is wrapped in a Debug directive (i.e. cases described by @mavasani, in the doc and in #3336).
